File: packages/core-ts/src/energy.ts
/**
 * Per-model inference energy estimation.
 *
 * Replaces the v0.1–v0.9 flat `ENERGY_KWH_PER_TASK = 0.0015` placeholder
 * with cited per-model Wh/token coefficients drawn from public research.
 *
 * The flat constant is preserved as the backwards-compatible fallback
 * for callers that have no model information (closure-based `defer`,
 * pre-v0.10 telemetry replays, etc.). Callers that provide a `model`
 * name plus optional input/output token counts get calibrated math.
 *
 * ## Sources
 *
 * - **Patterson et al. 2021.** "Carbon Emissions and Large Neural
 *   Network Training." arXiv:2104.10350. Per-model training energy
 *   for GPT-3 / T5 / Meena / GShard / Switch. PUE 1.1 reference.
 * - **Luccioni, Jernite, Strubell 2024.** "Power Hungry Processing:
 *   Watts Driving the Cost of AI Deployment." FAccT 2024.
 *   arXiv:2311.16863. The first systematic per-task inference
 *   energy measurement across 88 models and 10 task classes.
 *   This is the primary source for the open-weight coefficients below.
 * - **Hugging Face AI Energy Score** (2024–). Living benchmark of
 *   measured Wh/query across hundreds of open-weight models.
 *   https://huggingface.co/AIEnergyScore
 *
 * ## Closed-model caveat
 *
 * For closed models (Anthropic Claude family, OpenAI GPT-4 family,
 * Google Gemini family) the per-token coefficients are **estimated**
 * — inferred from public parameter-count disclosures (where any
 * exist) and scaled along the Luccioni curve. They may be off by
 * ±50%. The `source` field on each entry records the confidence
 * tier so downstream UIs can communicate uncertainty.
 *
 * ## Power Usage Effectiveness (PUE)
 *
 * The published coefficients reflect chip-level draw. To get
 * grid-level energy you multiply by data-centre PUE. ebb-ai uses
 * `DEFAULT_PUE = 1.15` (industry average for hyperscalers; Google
 * reports 1.10, AWS ~1.15, Azure ~1.18). Callers can override.
 */

import {
  DEFAULT_PUE as DEFAULT_PUE_DATA,
  ENERGY_SOURCES as ENERGY_SOURCES_DATA,
  LEGACY_KWH_PER_TASK as LEGACY_KWH_PER_TASK_DATA,
  MODEL_ENERGY_COEFFICIENTS as MODEL_ENERGY_COEFFICIENTS_DATA,
  MODEL_FAMILIES as MODEL_FAMILIES_DATA,
  TYPICAL_INPUT_TOKENS,
  TYPICAL_OUTPUT_TOKENS,
} from "./data/tables.generated.js";

/** Confidence tier for a coefficient entry. */
export type EnergySourceTier = "measured" | "estimated" | "fallback";

/**
 * How a coefficient was resolved for a given model id (v0.13+). This is
 * the receipt's `energyResolution` provenance — orthogonal to the
 * `EnergySourceTier` confidence of the number itself:
 *
 *   - `exact`          — the id matched a table key verbatim (case-folded).
 *   - `normalized`     — matched after stripping dated / provider / order
 *                        variance (see `normalizeModelName`).
 *   - `family-fallback`— the id is unknown but its family is known, so a
 *                        family representative's coefficients were used.
 *   - `default`        — fully unrecognized; the flat legacy constant.
 */
export type EnergyResolutionTier =
  | "exact"
  | "normalized"
  | "family-fallback"
  | "default";

/** Per-model inference energy coefficients (chip-level Wh, no PUE applied). */
export interface ModelEnergyCoefficients {
  /** Wh consumed per input (prompt) token. */
  whPerInputToken: number;
  /** Wh consumed per output (completion) token. */
  whPerOutputToken: number;
  /** Approximate total parameter count in billions (for context). */
  paramsB?: number;
  /** Provenance of the numbers. */
  source: EnergySourceTier;
}

/**
 * A family-fallback rule. When a model id is unknown but recognizably a
 * member of a family, the family's `representative` supplies coefficients.
 * Sourced from the JSON SSOT; a rule matches when every present condition
 * (`contains` / `prefix` / `regex`) holds against the normalized id.
 */
export interface ModelFamily {
  id: string;
  representative: string;
  contains?: readonly string[];
  prefix?: string;
  regex?: string;
}

/** The outcome of resolving a model id to coefficients + a provenance tier. */
export interface ResolvedModelEnergy {
  coeffs: ModelEnergyCoefficients;
  tier: EnergyResolutionTier;
}

/** Industry-average Power Usage Effectiveness for hyperscaler data centres. */
export const DEFAULT_PUE = DEFAULT_PUE_DATA;

/** Backwards-compatible flat estimate. Used when no model is provided. */
export const LEGACY_KWH_PER_TASK = LEGACY_KWH_PER_TASK_DATA;

/**
 * Per-model coefficient table. Keys are canonical lowercase names
 * (no version-date suffixes; see `normalizeModelName`). Sourced from the
 * JSON SSOT via the generated data module.
 */
export const MODEL_ENERGY_COEFFICIENTS = MODEL_ENERGY_COEFFICIENTS_DATA;

/** Ordered family-fallback rules (from the JSON SSOT). */
export const MODEL_FAMILIES = MODEL_FAMILIES_DATA;

/**
 * Citation metadata for the coefficient table. Exported so dashboards
 * and receipts can render attribution alongside numbers.
 */
export const ENERGY_SOURCES = ENERGY_SOURCES_DATA;

const FALLBACK_COEFFICIENTS: ModelEnergyCoefficients = {
  whPerInputToken: LEGACY_KWH_PER_TASK * 1000 / (TYPICAL_INPUT_TOKENS + TYPICAL_OUTPUT_TOKENS),
  whPerOutputToken: LEGACY_KWH_PER_TASK * 1000 / (TYPICAL_INPUT_TOKENS + TYPICAL_OUTPUT_TOKENS),
  source: "fallback",
};

/**
 * Strip provider prefixes, version-date suffixes and normalise punctuation
 * and word order so callers can pass `"claude-sonnet-4-5-20251022"`,
 * `"anthropic/claude-3.5-sonnet"`, `"gpt-4o-2024-11-20"` or
 * `"us.anthropic.claude-opus-4-7-v1:0"` and still hit the canonical entry.
 */
export function normalizeModelName(model: string): string {
  let name = model.trim().toLowerCase();
  // Strip a path-style provider prefix ("anthropic/…", "meta-llama/…").
  if (name.includes("/")) name = name.slice(name.lastIndexOf("/") + 1);
  // Strip a Bedrock region.vendor. prefix ("us.anthropic.…").
  name = name.replace(
    /^(?:us|eu|apac)\.(?:anthropic|meta|amazon|cohere|mistral|ai21|stability)\./,
    "",
  );
  // Strip a Bedrock trailing version tag (":0", ":1").
  name = name.replace(/:\d+$/, "");
  // Replace dots with dashes (`gpt-3.5-turbo` → `gpt-3-5-turbo`).
  name = name.replace(/\./g, "-");
  // Strip Anthropic / OpenAI dated suffix (-YYYYMMDD or -YYYY-MM-DD).
  name = name.replace(/-\d{4}-?\d{2}-?\d{2}$/, "");
  // Strip generic vendor suffixes (-latest, -preview, -v\d+, -\d{3,4}).
  name = name.replace(/-(latest|preview)$/, "");
  name = name.replace(/-v\d+$/, "");
  name = name.replace(/-\d{3,4}$/, "");
  // Canonicalize Claude word order: "claude-3-5-sonnet" → "claude-sonnet-3-5".
  const reordered = name.match(
    /^claude-(\d+(?:-\d+)*)-(opus|sonnet|haiku)(-.*)?$/,
  );
  if (reordered) {
    name = `claude-${reordered[2]}-${reordered[1]}${reordered[3] ?? ""}`;
  }
  return name;
}

/**
 * Return the coefficients of the family representative for a normalized
 * model id, or `undefined` if no family rule matches. The first matching
 * rule (in `MODEL_FAMILIES` order) wins.
 */
function familyRepresentative(
  normalized: string,
): ModelEnergyCoefficients | undefined {
  for (const fam of MODEL_FAMILIES) {
    if (fam.prefix !== undefined && !normalized.startsWith(fam.prefix)) continue;
    if (fam.contains && !fam.contains.every((t) => normalized.includes(t))) {
      continue;
    }
    if (fam.regex !== undefined && !new RegExp(fam.regex).test(normalized)) {
      continue;
    }
    return MODEL_ENERGY_COEFFICIENTS[fam.representative];
  }
  return undefined;
}

/**
 * Resolve a (possibly messy) model id to coefficients plus the provenance
 * tier describing *how* the match was made. See {@link EnergyResolutionTier}.
 */
export function resolveModelEnergy(model?: string): ResolvedModelEnergy {
  if (!model) return { coeffs: FALLBACK_COEFFICIENTS, tier: "default" };
  const exact = MODEL_ENERGY_COEFFICIENTS[model.trim().toLowerCase()];
  if (exact) return { coeffs: exact, tier: "exact" };
  const normalized = normalizeModelName(model);
  const norm = MODEL_ENERGY_COEFFICIENTS[normalized];
  if (norm) return { coeffs: norm, tier: "normalized" };
  const family = familyRepresentative(normalized);
  if (family) return { coeffs: family, tier: "family-fallback" };
  return { coeffs: FALLBACK_COEFFICIENTS, tier: "default" };
}

/** Look up coefficients for a (possibly suffixed) model name. */
export function lookupModelEnergy(model?: string): ModelEnergyCoefficients {
  return resolveModelEnergy(model).coeffs;
}

export interface EstimateEnergyOpts {
  /** Provider model identifier; e.g. "claude-sonnet-4-5", "gpt-4o". */
  model?: string;
  /** Prompt tokens. If omitted with a known model, `TYPICAL_INPUT_TOKENS` is used. */
  inputTokens?: number;
  /** Completion tokens. If omitted with a known model, `TYPICAL_OUTPUT_TOKENS` is used. */
  outputTokens?: number;
  /** Override PUE. Defaults to `DEFAULT_PUE` (1.15). */
  pue?: number;
}

/**
 * Estimate the grid-level energy (kWh) for an inference call.
 *
 * Resolution order:
 *   1. No arguments at all → `LEGACY_KWH_PER_TASK` (0.0015 kWh). The
 *      v0.1–v0.9 flat estimate, preserved bit-exactly for backwards
 *      compatibility with closure-based `defer` and replays of
 *      pre-v0.10 telemetry.
 *   2. Unknown model name → same fallback.
 *   3. Known model, no token counts → typical-task estimate
 *      (`TYPICAL_INPUT_TOKENS` + `TYPICAL_OUTPUT_TOKENS`) times the
 *      model's coefficients, times PUE.
 *   4. Known model + token counts → exact per-token math, times PUE.
 */
export function estimateEnergyKwh(opts: EstimateEnergyOpts = {}): number {
  const { model, inputTokens, outputTokens, pue = DEFAULT_PUE } = opts;

  // Path 1: backwards-compat zero-knowledge call.
  if (
    model === undefined &&
    inputTokens === undefined &&
    outputTokens === undefined
  ) {
    return LEGACY_KWH_PER_TASK;
  }

  const { coeffs, tier } = resolveModelEnergy(model);

  // Path 2: fully-unrecognized model with no token counts → legacy flat.
  // The fallback entry intentionally encodes the same total energy as
  // `LEGACY_KWH_PER_TASK` when summed across typical token counts; this
  // keeps "unknown model" identical to "no model" so dashboards don't
  // accidentally show a behavioural change for unmodelled traffic. A
  // *family-recognized* unknown (tier "family-fallback") instead uses the
  // family representative's coefficients — the §1.8 family fallback.
  if (tier === "default" && inputTokens === undefined && outputTokens === undefined) {
    return LEGACY_KWH_PER_TASK;
  }

  const ins = inputTokens ?? TYPICAL_INPUT_TOKENS;
  const outs = outputTokens ?? TYPICAL_OUTPUT_TOKENS;

  const chipWh = ins * coeffs.whPerInputToken + outs * coeffs.whPerOutputToken;
  const gridWh = chipWh * pue;
  return gridWh / 1000;
}

/**
 * Grams CO2-equivalent for one inference call at a given grid intensity.
 * Convenience helper used by the scheduler scoring path.
 *
 * @param gCo2PerKwh  Grid carbon intensity at the dispatch hour.
 * @param opts        Per-model + token args, same shape as `estimateEnergyKwh`.
 */
export function gramsForIntensity(
  gCo2PerKwh: number,
  opts: EstimateEnergyOpts = {},
): number {
  return estimateEnergyKwh(opts) * gCo2PerKwh;
}

File: packages/core-py/src/ebb_ai/energy.py
"""Per-model inference energy estimation (Python mirror of ``energy.ts``).

Replaces the v0.1-v0.5 flat ``ENERGY_KWH_PER_TASK = 0.0015`` placeholder
with cited per-model Wh/token coefficients drawn from public research.

The flat constant is preserved as the backwards-compatible fallback for
callers that have no model information (closure-based ``defer``, pre-v0.6
telemetry replays, etc.). Callers that pass a model name plus optional
input/output token counts get calibrated math.

Sources
-------
* **Patterson et al. 2021.** "Carbon Emissions and Large Neural Network
  Training." arXiv:2104.10350.
* **Luccioni, Jernite, Strubell 2024.** "Power Hungry Processing: Watts
  Driving the Cost of AI Deployment." FAccT 2024. arXiv:2311.16863.
* **Hugging Face AI Energy Score** (2024-).
  https://huggingface.co/AIEnergyScore

See ``packages/core-ts/src/energy.ts`` for the full design rationale and
closed-model caveat.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

from . import _data

EnergySourceTier = Literal["measured", "estimated", "fallback"]

#: How a coefficient was resolved for a given model id (v0.13+). Orthogonal
#: to :data:`EnergySourceTier` (the number's confidence). ``exact`` — the id
#: matched a table key verbatim; ``normalized`` — matched after stripping
#: dated / provider / word-order variance; ``family-fallback`` — unknown id,
#: known family, representative coefficients used; ``default`` — fully
#: unrecognized, flat legacy constant.
EnergyResolutionTier = Literal["exact", "normalized", "family-fallback", "default"]


@dataclass(frozen=True, slots=True)
class ModelEnergyCoefficients:
    """Per-model inference energy coefficients (chip-level Wh, no PUE)."""

    wh_per_input_token: float
    wh_per_output_token: float
    source: EnergySourceTier
    params_b: float | None = None


@dataclass(frozen=True, slots=True)
class ResolvedModelEnergy:
    """Coefficients plus the provenance tier describing how they matched."""

    coeffs: ModelEnergyCoefficients
    tier: EnergyResolutionTier


#: Industry-average Power Usage Effectiveness for hyperscaler data centres.
DEFAULT_PUE: float = _data.DEFAULT_PUE

#: Backwards-compatible flat estimate. Used when no model is provided.
LEGACY_KWH_PER_TASK: float = _data.LEGACY_KWH_PER_TASK

_TYPICAL_INPUT_TOKENS: int = _data.TYPICAL_INPUT_TOKENS
_TYPICAL_OUTPUT_TOKENS: int = _data.TYPICAL_OUTPUT_TOKENS

_FALLBACK = ModelEnergyCoefficients(
    wh_per_input_token=LEGACY_KWH_PER_TASK
    * 1000
    / (_TYPICAL_INPUT_TOKENS + _TYPICAL_OUTPUT_TOKENS),
    wh_per_output_token=LEGACY_KWH_PER_TASK
    * 1000
    / (_TYPICAL_INPUT_TOKENS + _TYPICAL_OUTPUT_TOKENS),
    source="fallback",
)


#: Per-model coefficient table. Keys are canonical lowercase names (no
#: dated suffixes; see ``normalize_model_name``). Built from the JSON SSOT
#: via the generated ``_data`` module.
MODEL_ENERGY_COEFFICIENTS: dict[str, ModelEnergyCoefficients] = {
    name: ModelEnergyCoefficients(
        wh_per_input_token=c["wh_per_input_token"],
        wh_per_output_token=c["wh_per_output_token"],
        source=c["source"],
        params_b=c["params_b"],
    )
    for name, c in _data.COEFFICIENTS.items()
}

#: Ordered family-fallback rules (from the JSON SSOT).
MODEL_FAMILIES: list[dict[str, object]] = _data.FAMILIES

#: Citation metadata for the coefficient table.
ENERGY_SOURCES: dict[str, dict[str, str]] = _data.ENERGY_SOURCES


_DATED_SUFFIX = re.compile(r"-\d{4}-?\d{2}-?\d{2}$")
_LATEST_PREVIEW = re.compile(r"-(latest|preview)$")
_VERSION_SUFFIX = re.compile(r"-v\d+$")
_NUMERIC_SUFFIX = re.compile(r"-\d{3,4}$")
_PROVIDER_DOTTED = re.compile(
    r"^(?:us|eu|apac)\.(?:anthropic|meta|amazon|cohere|mistral|ai21|stability)\."
)
_BEDROCK_TAG = re.compile(r":\d+$")
_CLAUDE_REORDER = re.compile(r"^claude-(\d+(?:-\d+)*)-(opus|sonnet|haiku)(-.*)?$")


def normalize_model_name(model: str) -> str:
    """Strip provider prefixes, version-date suffixes and normalise
    punctuation and word order.

    Callers can pass ``"claude-sonnet-4-5-20251022"``,
    ``"anthropic/claude-3.5-sonnet"``, ``"gpt-4o-2024-11-20"`` or
    ``"us.anthropic.claude-opus-4-7-v1:0"`` and still hit the canonical entry.
    """
    name = model.strip().lower()
    if "/" in name:
        name = name.rsplit("/", 1)[1]
    name = _PROVIDER_DOTTED.sub("", name)
    name = _BEDROCK_TAG.sub("", name)
    name = name.replace(".", "-")
    name = _DATED_SUFFIX.sub("", name)
    name = _LATEST_PREVIEW.sub("", name)
    name = _VERSION_SUFFIX.sub("", name)
    name = _NUMERIC_SUFFIX.sub("", name)
    m = _CLAUDE_REORDER.match(name)
    if m:
        name = f"claude-{m.group(2)}-{m.group(1)}{m.group(3) or ''}"
    return name


def _family_representative(normalized: str) -> ModelEnergyCoefficients | None:
    """Coefficients of the first family rule that matches, else ``None``."""
    for fam in MODEL_FAMILIES:
        prefix = fam.get("prefix")
        if prefix is not None and not normalized.startswith(prefix):
            continue
        contains = fam.get("contains")
        if contains is not None and not all(t in normalized for t in contains):
            continue
        regex = fam.get("regex")
        if regex is not None and re.search(regex, normalized) is None:
            continue
        return MODEL_ENERGY_COEFFICIENTS[fam["representative"]]
    return None


def resolve_model_energy(model: str | None = None) -> ResolvedModelEnergy:
    """Resolve a (possibly messy) model id to coefficients + provenance tier.

    See :data:`EnergyResolutionTier`.
    """
    if not model:
        return ResolvedModelEnergy(_FALLBACK, "default")
    exact = MODEL_ENERGY_COEFFICIENTS.get(model.strip().lower())
    if exact is not None:
        return ResolvedModelEnergy(exact, "exact")
    normalized = normalize_model_name(model)
    norm = MODEL_ENERGY_COEFFICIENTS.get(normalized)
    if norm is not None:
        return ResolvedModelEnergy(norm, "normalized")
    family = _family_representative(normalized)
    if family is not None:
        return ResolvedModelEnergy(family, "family-fallback")
    return ResolvedModelEnergy(_FALLBACK, "default")


def lookup_model_energy(model: str | None = None) -> ModelEnergyCoefficients:
    """Look up coefficients for a (possibly suffixed) model name."""
    return resolve_model_energy(model).coeffs


def estimate_energy_kwh(
    *,
    model: str | None = None,
    input_tokens: int | None = None,
    output_tokens: int | None = None,
    pue: float = DEFAULT_PUE,
) -> float:
    """Estimate grid-level energy (kWh) for an inference call.

    Resolution order:
        1. No arguments → :data:`LEGACY_KWH_PER_TASK` (0.0015 kWh).
           Backwards-compatible behaviour for closure-based ``defer``.
        2. Unknown model name (no tokens) → same fallback.
        3. Known model, no token counts → typical-task estimate
           (500 + 500 tokens) times the model's coefficients, times PUE.
        4. Known model + token counts → exact per-token math, times PUE.
    """
    if model is None and input_tokens is None and output_tokens is None:
        return LEGACY_KWH_PER_TASK

    resolved = resolve_model_energy(model)
    coeffs = resolved.coeffs

    # Only fully-unrecognized models with no token counts fall back to the
    # flat legacy constant. A family-recognized unknown (tier
    # ``family-fallback``) instead uses the family representative's
    # coefficients — the §1.8 family fallback.
    if (
        resolved.tier == "default"
        and input_tokens is None
        and output_tokens is None
    ):
        return LEGACY_KWH_PER_TASK

    ins = input_tokens if input_tokens is not None else _TYPICAL_INPUT_TOKENS
    outs = output_tokens if output_tokens is not None else _TYPICAL_OUTPUT_TOKENS

    chip_wh = ins * coeffs.wh_per_input_token + outs * coeffs.wh_per_output_token
    grid_wh = chip_wh * pue
    return grid_wh / 1000.0


def grams_for_intensity(
    g_co2_per_kwh: float,
    *,
    model: str | None = None,
    input_tokens: int | None = None,
    output_tokens: int | None = None,
    pue: float = DEFAULT_PUE,
) -> float:
    """Grams CO2-equivalent for one inference call at a given grid intensity."""
    kwh = estimate_energy_kwh(
        model=model,
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        pue=pue,
    )
    return kwh * g_co2_per_kwh


__all__ = [
    "DEFAULT_PUE",
    "ENERGY_SOURCES",
    "LEGACY_KWH_PER_TASK",
    "MODEL_ENERGY_COEFFICIENTS",
    "MODEL_FAMILIES",
    "EnergyResolutionTier",
    "EnergySourceTier",
    "ModelEnergyCoefficients",
    "ResolvedModelEnergy",
    "estimate_energy_kwh",
    "grams_for_intensity",
    "lookup_model_energy",
    "normalize_model_name",
    "resolve_model_energy",
]
